### ISSUE TITLE
Firewall backend redesign: Strategy pattern for iptables/nftables/Docker environments

### ISSUE THREAD

**Issue by lenny-ts** [lenny-ts/caddy-analyzer#80]
## Problem

The current IP blocking implementation uses a single hardcoded approach:

```
iptables -A INPUT -j CADDY_ANALYZER
iptables -A CADDY_ANALYZER -s <IP> -m comment --comment "caddy-analyzer" -j DROP
```

This fails in multiple real-world scenarios:

| Scenario | Why it fails |
|----------|-------------|
| **Docker containers** | Traffic goes through `FORWARD`, not `INPUT`. DROP rules in INPUT are never evaluated. |
| **nftables backend** | `iptables -C` (check) can return false positives on nftables translations, causing `BlockIP` to skip adding rules. |
| **Reverse proxy / CDN** | iptables sees the proxy IP, not the client IP. Blocking the wrong address. |
| **Rules appended (`-A`)** | If an early `ACCEPT` rule exists in INPUT, the CADDY_ANALYZER jump is never reached. |
| **loadState on restart** | iptables rules lost after reboot, but `g.blocked` map is restored — IPs appear blocked in-memory but firewall is open. |
| **IPv6 ignored** | If ip6tables chain doesn't exist or jump is missing, IPv6 rules silently fail. |

The root cause is architectural: the `Blocker` interface exists but has only one implementation (`iptablesBlocker`), and the chain strategy (`INPUT`) is hardcoded.

### Architecture (Hexagonal)

```
┌─────────────────────────────────────────────────────────┐
│                     GUARD (Domain)                      │
│                                                         │
│  block() ──→ Blocker (Port) ──→ [Strategy selected]     │
│                                  at startup             │
├─────────────────────────────────────────────────────────┤
│                  FIREWALL ADAPTERS                       │
│                                                         │
│  ┌──────────┐  ┌──────────┐  ┌──────────┐  ┌────────┐ │
│  │ iptables │  │ nftables │  │  docker  │  │ hybrid │ │
│  │  INPUT   │  │  nft set │  │ DOCKER-  │  │ auto   │ │
│  │          │  │          │  │ USER     │  │        │ │
│  └──────────┘  └──────────┘  └──────────┘  └────────┘ │
└─────────────────────────────────────────────────────────┘
```

### 1. FirewallBackend Interface (Port)

```go
// pkg/guard/firewall/firewall.go

package firewall

import "time"

// Backend is the port that the guard domain uses to block/unblock IPs.
// Implementations are adapters that translate to the actual firewall mechanism.
type Backend interface {
    // Block adds a DROP rule for ip. Idempotent.
    Block(ip string) error

    // Unblock removes the DROP rule for ip.
    Unblock(ip string) error

    // ListBlocked returns all IPs currently blocked by this tool.
    ListBlocked() ([]string, error)

    // EnsureChain creates the dedicated chain and jump rule if missing.
    // Called once at startup.
    EnsureChain() error

    // Validate checks that the chain and jump rule are correctly positioned.
    // Returns error if the firewall is not actually effective.
    Validate() error

    // Name returns the backend type for logging (e.g. "iptables", "nftables", "docker").
    Name() string
}
```

#### a. iptablesBackend (bare metal / legacy)

```go
// pkg/guard/firewall/iptables.go

type iptablesBackend struct {
    bin     string   // "iptables" or "ip6tables"
    chain   string   // "CADDY_ANALYZER"
    comment string   // "caddy-analyzer"
}

func (b *iptablesBackend) EnsureChain() error {
    // Create chain if missing
    // Insert jump at TOP of INPUT: -I INPUT 1 -j CADDY_ANALYZER
    // NOT append (-A) which goes after existing ACCEPT rules
}

func (b *iptablesBackend) Block(ip string) error {
    // 1. Check if rule exists (-C)
    // 2. If not, insert at top of chain (-I, not -A)
    // 3. Verify rule exists after insertion with -S
    // 4. If verification fails, return error (don't silently succeed)
}

func (b *iptablesBackend) Validate() error {
    // 1. Check chain exists: iptables -L CADDY_ANALYZER -n
    // 2. Check jump exists: iptables -C INPUT -j CADDY_ANALYZER
    // 3. Check jump position: iptables -S INPUT | head -5
    // 4. Verify packet counters are incrementing (rules are being hit)
}
```

#### b. dockerBackend (Docker environment)

```go
// pkg/guard/firewall/docker.go

type dockerBackend struct {
    bin     string
    chain   string
    comment string
}

func (b *dockerBackend) EnsureChain() error {
    // 1. Create CADDY_ANALYZER chain
    // 2. Insert jump into DOCKER-USER chain (not INPUT!)
    //    iptables -I DOCKER-USER -j CADDY_ANALYZER
    // 3. Verify DOCKER-USER chain exists
}

func (b *dockerBackend) Block(ip string) error {
    // Same as iptablesBackend but targets DOCKER-USER chain
}

func (b *dockerBackend) Validate() error {
    // 1. Check DOCKER-USER chain exists
    // 2. Check jump from DOCKER-USER to CADDY_ANALYZER
    // 3. Warn if Caddy container uses --network host (rules in INPUT still needed)
}
```

#### c. nftablesBackend (nftables-native)

```go
// pkg/guard/firewall/nftables.go

type nftablesBackend struct {
    chain   string
    set     string   // nft set name for blocked IPs
}

func (b *nftablesBackend) Block(ip string) error {
    // 1. Add IP to nft set: nft add element inet filter caddy_analyzer { <IP> }
    // 2. Verify element exists: nft list set inet filter caddy_analyzer
    // 3. nftables sets are O(1) lookup — better performance than iptables linear scan
}

func (b *nftablesBackend) Validate() error {
    // 1. Check set exists: nft list sets | grep caddy_analyzer
    // 2. Check rule referencing set exists
    // 3. Verify element count matches expected
}
```

#### d. hybridBackend (auto-detect and apply to all relevant chains)

```go
// pkg/guard/firewall/hybrid.go

type hybridBackend struct {
    backends []Backend  // one per relevant chain
}

func (b *hybridBackend) Block(ip string) error {
    // Block on ALL backends (INPUT + DOCKER-USER + FORWARD if needed)
    // Returns error only if ALL backends fail
}

func (b *hybridBackend) Validate() error {
    // Validate each backend, collect warnings
}
```

### 3. Environment Detection (Composition Root)

```go
// pkg/guard/firewall/detect.go

func Detect() Backend {
    // Step 1: Check if Docker is running
    if isDockerRunning() {
        // Step 2: Check if Caddy is in a Docker container
        if isCaddyInDocker() {
            // Step 3: Check if container uses host networking
            if isHostNetwork() {
                // Host network: INPUT chain is correct
                return newiptablesBackend("iptables")
            }
            // Bridge networking: DOCKER-USER chain is needed
            return newDockerBackend()
        }
        // Docker on host but Caddy is bare metal: use INPUT
        return newiptablesBackend("iptables")
    }

    // Step 4: Check if nftables backend
    if isNftablesBackend() {
        return newNftablesBackend()
    }

    // Step 5: Default to iptables with INPUT chain
    return newiptablesBackend("iptables")
}

func isDockerRunning() bool {
    // Check if docker0 interface exists or dockerd is running
    // os.Stat("/var/run/docker.sock") == nil
}

func isCaddyInDocker() bool {
    // Check if we're inside a container:
    // - /.dockerenv exists
    // - /proc/1/cgroup contains "docker"
    // Or check if Caddy's PID is in a different namespace
}

func isHostNetwork() bool {
    // Check if container uses --network host
    // docker inspect <container> --format '{{.HostConfig.NetworkMode}}'
}

func isNftablesBackend() bool {
    // iptables --version | grep nf_tables
    // Or check if nft command exists and iptables is a symlink
}
```

#### a. loadState must recreate iptables rules

```go
// pkg/guard/guard.go — loadState()

case e.When.IsZero():
    g.blocked[e.IP] = true
    if err := g.blocker.Block(e.IP); err != nil && g.cfg.OnError != nil {
        g.cfg.OnError(fmt.Errorf("restore permanent block %s: %w", e.IP, err))
    }

default:
    g.blocked[e.IP] = true
    if err := g.blocker.Block(e.IP); err != nil && g.cfg.OnError != nil {
        g.cfg.OnError(fmt.Errorf("restore block %s: %w", e.IP, err))
    }
    g.expiries[e.IP] = e.When
    g.initial = append(g.initial, expiryEntry{ip: e.IP, when: e.When})
```

#### b. Insert rules at top, not append

```go
// BEFORE (broken):
args := []string{chainName, "-s", ip, "-m", "comment", "--comment", commentMarker, "-j", "DROP"}
return runCmd(context.Background(), bin, append([]string{"-A"}, args...)...)

// AFTER (correct):
args := []string{chainName, "-s", ip, "-m", "comment", "--comment", commentMarker, "-j", "DROP"}
return runCmd(context.Background(), bin, append([]string{"-I"}, args...)...)
```

#### c. Verify rules after insertion

```go
func (b *iptablesBackend) Block(ip string) error {
    // ... add rule ...
    
    // VERIFY: check rule actually exists
    args := []string{chainName, "-s", ip, "-m", "comment", "--comment", commentMarker, "-j", "DROP"}
    if err := runCmd(ctx, b.bin, append([]string{"-C"}, args...)...); err != nil {
        return fmt.Errorf("rule verification failed for %s: %w", ip, err)
    }
    return nil
}
```

#### d. Startup validation

```go
// In Run() or New():
if err := g.blocker.EnsureChain(); err != nil {
    return fmt.Errorf("firewall setup: %w", err)
}
if err := g.blocker.Validate(); err != nil {
    g.cfg.OnError(fmt.Errorf("firewall validation: %w", err))
    // Don't fail — degraded protection is better than no protection
    // But log prominently
}
```

### 5. File Structure

```
pkg/guard/
├── firewall/
│   ├── firewall.go        # Backend interface (port)
│   ├── iptables.go        # iptables adapter (INPUT chain)
│   ├── docker.go          # Docker adapter (DOCKER-USER chain)
│   ├── nftables.go        # nftables adapter (nft sets)
│   ├── hybrid.go          # Multi-chain adapter
│   ├── detect.go          # Environment detection
│   ├── detect_test.go     # Detection tests
│   ├── iptables_test.go   # iptables adapter tests
│   ├── docker_test.go     # Docker adapter tests
│   └── nftables_test.go   # nftables adapter tests
├── guard.go               # Uses firewall.Backend (unchanged interface)
├── guard_test.go          # Uses fakeBlocker (unchanged)
└── state.go               # Unchanged
```

### 6. Migration Path

1. **Phase 1**: Fix `loadState` to recreate iptables rules (immediate fix)
2. **Phase 2**: Extract `firewall.Backend` interface, move existing code to `iptablesBackend`
3. **Phase 3**: Add environment detection, create `dockerBackend` and `nftablesBackend`
4. **Phase 4**: Add `Validate()` calls at startup and after block operations
5. **Phase 5**: Add `--firewall-backend` flag for manual override (auto-detect default)

### 7. CLI Flags (New)

```
--firewall-backend string    Force firewall backend: auto, iptables, docker, nftables (default: auto)
--firewall-chain string      Override chain name (default: auto-detect)
--firewall-verify            Verify rules after each block (default: true)
```

### 8. Diagnostic Command

Add `caddy-analyze guard --check` that validates the full chain:

```bash
$ caddy-analyze guard --check

Firewall diagnostics:
  Backend:        docker (auto-detected)
  Chain:          CADDY_ANALYZER
  Jump chain:     DOCKER-USER
  Jump position:  1 (top)
  
  Chain exists:   ✓
  Jump exists:    ✓
  Jump position:  ✓ (rule 1 of DOCKER-USER)
  
  Blocked IPs:    47
  iptables rules: 47 (match ✓)
  State file:     47 entries (match ✓)
  
  Packet counters: 1,234 hits in last hour ✓
  
  IPv6:           ip6tables chain exists ✓
  
  Warnings:       none
```

## Testing Strategy

1. **Unit tests**: Use `fakeBlocker` (existing) for guard logic tests
2. **Integration tests**: Mock iptables/nftables commands for adapter tests
3. **Docker tests**: Spin up a Docker container, verify DOCKER-USER chain rules
4. **End-to-end**: Deploy on a server with Docker, verify blocked IPs cannot reach Caddy

## References

- [Docker DOCKER-USER chain](https://docs.docker.com/engine/network/packet-filtering-firewalls/#manage-iptables-for-docker)
- [nftables wiki](https://wiki.nftables.org/)
- [iptables vs nftables](https://www.redhat.com/en/blog/linux-iptables-nftables)
- Clean Architecture (Robert C. Martin) — Port and Adapter pattern
- pfSense firewall rule processing order (first match wins)
